io/src/lib.rs

```diff
@@ -0,0 +1,506 @@
+        NotConnected,
+        AddrInUse,
+        AddrNotAvailable,
+        BrokenPipe,
+        AlreadyExists,
+        WouldBlock,
+        InvalidInput,
+        InvalidData,
+        TimedOut,
+        WriteZero,
+        Interrupted,
+        UnexpectedEof,
+        // Note: Any time we bump the MSRV any new error kinds should be added here!
+        Other
+    );
+
+    pub type Result<T> = core::result::Result<T, Error>;
+
+    /// A generic trait describing an input stream. See [`std::io::Read`] for more info.
+    pub trait Read {
+        fn read(&mut self, buf: &mut [u8]) -> Result<usize>;
+        #[inline]
+        fn read_exact(&mut self, mut buf: &mut [u8]) -> Result<()> {
+            while !buf.is_empty() {
+                match self.read(buf) {
+                    Ok(0) => return Err(ErrorKind::UnexpectedEof.into()),
+                    Ok(len) => buf = &mut buf[len..],
+                    Err(e) if e.kind() == ErrorKind::Interrupted => {}
+                    Err(e) => return Err(e),
+                }
+            }
+            Ok(())
+        }
+        #[inline]
+        fn take(&mut self, limit: u64) -> Take<Self> {
+            Take { reader: self, remaining: limit }
+        }
+        #[inline]
+        #[cfg(any(feature = "alloc", feature = "std"))]
+        fn read_to_end(&mut self, buf: &mut alloc::vec::Vec<u8>) -> Result<usize> {
+            let mut bytes_read = 0;
+            let mut read_buf = [0u8; 512];
+            loop {
+                match self.read(&mut read_buf) {
+                    Ok(0) => break,
+                    Ok(n) => {
+                        buf.extend_from_slice(&read_buf[0..n]);
+                        bytes_read += n;
+                    }
+                    Err(ref e) if e.kind() == ErrorKind::Interrupted => {}
+                    Err(e) => return Err(e),
+                }
+            }
+            Ok(bytes_read)
+        }
+    }
+
+    pub struct Take<'a, R: Read + ?Sized> {
+        reader: &'a mut R,
+        remaining: u64,
+    }
+    impl<'a, R: Read + ?Sized> Read for Take<'a, R> {
+        #[inline]
+        fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+            let len = core::cmp::min(buf.len(), self.remaining.try_into().unwrap_or(buf.len()));
+            let read = self.reader.read(&mut buf[..len])?;
+            self.remaining -= read.try_into().unwrap_or(self.remaining);
+            Ok(read)
+        }
+    }
+
+    #[cfg(feature = "std")]
+    impl<R: std::io::Read> Read for R {
+        #[inline]
+        fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+            Ok(<R as std::io::Read>::read(self, buf)?)
+        }
+    }
+
+    #[cfg(all(feature = "core2", not(feature = "std")))]
+    impl<R: core2::io::Read> Read for R {
+        #[inline]
+        fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+            Ok(<R as core2::io::Read>::read(self, buf)?)
+        }
+    }
+
+    pub struct Cursor<T> {
```

Comment:
The only purpose of `Cursor` is to provide `Seek` which we don't have here, so this shouldn't be needed. Reading alone can be done from slices directly.

Reply:
When you read from a slice it doesn't update the read pointer (there is no room for a read pointer) so you can't do repeated reads. This is what `Cursor` is for, far more often than `Seek`.

Reply:
It does.

Reply:
I remember seeing some hack related to cursor anyway, so maybe just get rid of it? At least at that place. Anyway, not a blocker for me.